File: Hashing2.java

// ===========================================Sub Array Sum================================================
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
//  - Sum of Sub array = k if the distance between two points in array is K. 
//  - Calculate running sum for the entire array and if running_sum - k is already available, then thats one sub array to be counted
// Your code here along with comments explaining your approach


class Solution {
    public int subarraySum(int[] nums, int k) {
        if(nums == null || nums.length ==0)
            return 0;
        
        int running_sum =0;
        int count =0;
        Map<Integer, Integer> hm = new HashMap<Integer, Integer>();
        hm.put(0,1);
        for(int i =0; i< nums.length; i++){
            running_sum += nums[i];
            int diff = running_sum - k;
            if(hm.containsKey(diff))
                count += hm.get(diff);
                // hm.put(running_sum,1);
            hm.put(running_sum, hm.getOrDefault(running_sum,0)+1);
            //     if(hm.containsKey(running_sum)){
            //     hm.put(running_sum,hm.get(running_sum)+1);
            // }else{
            //     hm.put(running_sum,1);
            // }
        }
        return count;
    }
}


// =================================Longest Palaindrome=================================

// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :No
// Your code here along with comments explaining your approach
class Solution {
    public int longestPalindrome(String s) {
        
        HashSet<Character> hs = new HashSet<Character>();
        int count = 0;
        for(int i = 0; i< s.length(); i++){
            if(hs.contains(s.charAt(i))){
                hs.remove(s.charAt(i));
                count++;
            }else{
                hs.add(s.charAt(i));
            }
        }
        if(hs.isEmpty())
                return count*2;
            else
                return count*2+1;
        
    }
}

// ======================= No of Continous Array in binary Array=====================

// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :No
// Your code here along with comments explaining your approach



class Solution {
    public int findMaxLength(int[] nums) {
        if(nums ==null || nums.length == 0)
            return -1;
        
        HashMap<Integer, Integer> hashMap = new HashMap<Integer, Integer>(); // Hash with Running sum and index of first occurence
        int len = nums.length;
        int max=0, count = 0;
        hashMap.put(0,-1); 
        for(int i = 0; i< len; i++){
            count += nums[i] ==0? -1:1;
            if(hashMap.containsKey(count)){
                max = Math.max(max,i - hashMap.get(count)); // If the running sum is available then get the previous index(value) and 
                                                           //  check if the difference of current index to get maximum size of sub array                                                                     
            }else{
                hashMap.put(count, i);  // Add the running sum first time with its initial occurence index
            }
            
        }
        return max;
    }
}

File: Hashing-2.py
#Problem 1
#Time Complexity: O(n)
#Space Complexity: O(n)
#Did this code successfully run on leetcode: yes
'''
Main idea:
We traverse the array to have the total sum until i for each i.
The thing to notice here is that if sum[j]-sum[i]=k, it means that
nums[i]+nums[i+1]+...+nums[j]=k, since up to i the elements are the same.
So when traversing the array, we can store the sum as key in the hashmap and
increment the count of that sum as the value for the hashmap.We can
check if in that hashmap how many times we had stored sum[i]-k,  let's call them
j (their index). It would mean that nums[j]+nums[j+1]+...+nums[i] is k. We increment
a count of every such occurences and return it at the end.
'''

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        if not nums:
            return 0
        h={}
        sum_i=0
        equals_k=0
        for i in range(len(nums)):
            sum_i+=nums[i]
            if sum_i-k in h:
                #this block should go before because if target 0, it will be counted
                # one time more
                #if sum_i-k is in h, it means there's a j <i such that sum_i-sum_j is k
                #meaning that the indices from j to i sum to k
                equals_k+=h[sum_i-k]
            if sum_i==k:
                #It is possible that the cumulative sum is already k
                equals_k+=1
            if sum_i in h:
                h[sum_i]+=1
            else:
                h[sum_i]=1
        return equals_k



#Problem 2
#Time Complexity: O(n)
#Space Complexity: O(n)
#Did this code successfully run on leetcode: yes
'''
Main idea:
if 0 we -1 the sum if 1 we +1 the sum. If the count is zero it means it's the same.
if the count is the same at two points, it means that the indices in between (i to j)
have the same number of ones and zero, because it means that it added 0
from i to j, meaning same number of 0 and 1. Also we can store the first occurence of each
sum, since we want the longest continuous subarray. So when we encounter a sum,
we check it it's in the hash, and do that index-same_occurence_first_index to get the length.
We compare it to the max length up until now and update. Similar to problem 1.
'''

class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        #idea: if sum=0, has property
        #edge case
        if not nums:
            return 0
        sum_i=0
        max_sub=0
        h={}
        for i in range(len(nums)):
            if nums[i]==1:
                sum_i+=1
            else:
                sum_i-=1
            if sum_i==0:
                #if the whole array until now has a count of 0
                #obviously it's the max length until now
                max_sub=i+1
                continue
            if sum_i in h:
                # i-h[sum_i] is the length of subarray that has the property, since we exclude the first index
                max_sub=max(i-h[sum_i],max_sub)
            else:
                h[sum_i]=i
        return max_sub


#Problem 3
#Time Complexity: O(n)
#Space Complexity: O(n)
#Did this code successfully run on leetcode: yes
'''
Main idea:
Keep a count of each letters, every pair of 2 can be added to my palindrome word
also keep track of letters to determine if we can add a letter to the center
to increase count
'''
class Solution:
    def longestPalindrome(self, s: str) -> int:
        #edge case
        if not s:
            return 0
        counts={}
        letters_left=len(s)
        max_l=0
        for c in s:
            if c in counts:
                counts[c]+=1
            else:
                counts[c]=1
        for c in counts:
            #if we have two of them, we can add the two in the palindrome
            n=2*(counts[c]//2)
            max_l+=n
            letters_left-=n
            #keep track of letters left to see if we can add on in the middle at the end

        if letters_left:
            return max_l+1
        return max_l
